blog/features/wasm-native-stable-memory.md

```diff
@@ -0,0 +1,94 @@
+ 
+## How Wasm-native Stable Memory works
+
+Fortunately the new [multiple memories](https://github.com/WebAssembly/multi-memory/blob/master/proposals/multi-memory/Overview.md) Wasm feature allows a Wasm module to directly address multiple byte arrays (called “memories” in the Wasm spec). We can use this to modify calls to the stable memory APIs to directly read or write from a secondary Wasm memory.
+
+The following diagram shows how stable memory accesses will work under Wasm-native stable memory:
+
+![Wasm-native stable memory diagram](/img/blog/wasm-native-stable-memory-diagram-new.png)
+
+## Performance Gains on Sample App
+
+Here we take an example dapp using the `ic-stable-structures` library and show how its performance improves after the Wasm-native stable change. Let’s start with the [basic example](https://github.com/dfinity/stable-structures/tree/main/examples/src/basic_example) from `ic-stable-structures` repo which implements a key-value store. The dapp starts by setting up a `StableBTreeMap` to store data:
+
+```rust
+type Memory = VirtualMemory<DefaultMemoryImpl>;
+
+thread_local! {
+    // The memory manager is used for simulating multiple memories. Given a `MemoryId` it can
+    // return a memory that can be used by stable structures.
+    static MEMORY_MANAGER: RefCell<MemoryManager<DefaultMemoryImpl>> =
+        RefCell::new(MemoryManager::init(DefaultMemoryImpl::default()));
+
+    // Initialize a `StableBTreeMap` with `MemoryId(0)`.
+    static MAP: RefCell<StableBTreeMap<u128, u128, Memory>> = RefCell::new(
+        StableBTreeMap::init(
+            MEMORY_MANAGER.with(|m| m.borrow().get(MemoryId::new(0))),
+        )
+    );
+}
+```
+
+We then enhance the basic example by adding an API to bulk insert many key-value pairs:
+
+```rust
+#[ic_cdk_macros::update]
+fn insert_many(entries: Vec<(u128, u128)>) {
+    MAP.with(|p| {
+        let mut map = p.borrow_mut();
+        for (k, v) in entries.into_iter() {
+            map.insert(k, v);
+        }
+    })
+}
+```
+
+After deploying this dapp to the IC, we then executed multiple messages that bulk insert many key-value pairs (close to maxing out the ingress message size limit). The following graphs show the execution time for each of these messages.
+
+First without Wasm-native stable memory:
+
+![Metrics without Wasm-native stable memory](/img/blog/wasm-native-stable-memory-execution-metrics-no-wnsm.png)
+
+Then with Wasm-native stable memory:
+
+![Metrics with Wasm-native stable memory](/img/blog/wasm-native-stable-memory-execution-metrics-wnsm.png)
+
+So we can see that this dapp gets a 1.5x speedup from Wasm-native stable memory.
+
+## Production Examples
+
+If you’re interested in seeing some architecture of more complicated dapps that are using stable memory, the [bitcoin canister](https://github.com/dfinity/bitcoin-canister) and the [internet identity canister](https://github.com/dfinity/internet-identity/tree/main/src/internet_identity) are good examples.
```

Comment:
Could you add a 'wrap up' or 'conclusion' section after this that closes out the blog post? As it stands, it feels like it ends a bit abruptly.

Reply:
I just added a short conclusion section.